[docs] Create page on inference servers with transformers backend (#39550)
huggingface/transformers: docs/source/en/transformers_as_backend.md

```

Add `--trust-remote_code` to the command to serve a custom Transformers model.

```docker
docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:latest --model-id <CUSTOM_MODEL_ID> --trust-remote-code
```

## Making Your Model Compatible Once and For All Backends

To make your custom model work out of the box with backends like vLLM and SGLang, it needs to follow some conventions, mainly to ensure smooth integration and optimized inference.

### General Requirements

For a model to be supported via the Transformers backend:

1. It must be Transformers-compatible following model guidelines. That means the model has to be supported in the core library  following [these rules](https://huggingface.co/docs/transformers/en/add_new_model) or contain custom code in the Hub following [custom code sharing guidelines](https://huggingface.co/docs/transformers/en/custom_models) in the Hub. Make sure that the model has a valid `config.json` in its directory and a valid `auto_map` field pointing to the model class in the config.

> 2. The model's attention module needs to be backend configurable to benefit from performance features of various inference servers. For that the model needs to support the new [AttentionInterface](https://huggingface.co/docs/transformers/en/attention_interface) which allows anyone to register their custom and optimized attention functions to be used in the model. All you have to do is to use `ALL_ATTENTION_FUNCTIONS` when defining the attention layer and propagate `**kwargs` all the way through your base `MyModel` class to the attention layers. Finally don't forget to set `_supports_attention_backend = True` in you `MyPreTrainedModel` class. Expand the below section for an example pseudo-code.

Reviewer: ```suggestion
2. A model needs to support the [AttentionInterface](./attention_interface) class to allow custom and optimized attention functions. This is important for enabling the performance features of the different inference servers.

   Use `ALL_ATTENTION_FUNCTIONS` when defining the attention layer and propagate `**kwargs** from the base `MyModel` class to the attention layers. Set `_supports_attention_backend` to `True` in [`PreTrainedModel`]. Expand the code below for an example.
```

<details>
<summary>modeling_my_model.py</summary>

```python

from transformers import PreTrainedModel
from torch import nn

class MyAttention(nn.Module):

    def forward(self, hidden_states, **kwargs):
        ...
        attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
        attn_output, attn_weights = attention_interface(
            self,
            query_states,
            key_states,
            value_states,
            **kwargs,
        )
        ...

class MyModel(PreTrainedModel):
    _supports_attention_backend = True
```

</details>

> 3. Optionally, if you want the model to support tensor parallel and/or pipeline parallel features, you can add the following keys in the config file: 

Reviewer: ```suggestion
3. This step is optional, but if you want to support tensor parallel and/or pipeline parallel features, add the following keys to the config.
```

>     * `base_model_tp_plan` for [tensor parallelism](https://huggingface.co/docs/transformers/perf_infer_gpu_multi) - a dict that maps fully qualified layer name patterns to tensor parallel styles (currently only "colwise" and "rowwise" are supported).

Reviewer: ```suggestion
    * `base_model_tp_plan` enables [tensor parallelism](./perf_infer_gpu_multi) by mapping fully qualified layer name patterns to tensor parallel styles. Only the `"colwise"` and `"rowwise"` partitioning strategies are currently supported.
```

>     * `base_model_pp_plan` for pipeline parallelism - a dict that maps direct child layer names to tuples of lists of strs.The list in the first element of the tuple contains the names of the input arguments. The list in the last element of the tuple contains the names of the variables the layer outputs to in your modeling code

Reviewer: ```suggestion
    * `base_model_pp_plan` enables pipeline parallelism by mapping direct child layer names to tuples of lists of strings. The list in the first element of the tuple contains the names of the input arguments. The list in the last element of the tuple contains the names of the variables the layer outputs to in the modeling code.
 
 Expand the code below for an example.
```

<details>
<summary>configuration_my_model.py</summary>

```python

from transformers import PretrainedConfig

class MyConfig(PretrainedConfig):
    base_model_tp_plan = {
        "layers.*.self_attn.k_proj": "colwise",
        "layers.*.self_attn.v_proj": "colwise",
        "layers.*.self_attn.o_proj": "rowwise",
        "layers.*.mlp.gate_proj": "colwise",
        "layers.*.mlp.up_proj": "colwise",
        "layers.*.mlp.down_proj": "rowwise",
    }
    base_model_pp_plan = {
        "embed_tokens": (["input_ids"], ["inputs_embeds"]),
        "layers": (["hidden_states", "attention_mask"], ["hidden_states"]),
        "norm": (["hidden_states"], ["hidden_states"]),
    }
```
</details>

> ### Multimodal Requirements

Reviewer: ```suggestion
### Multimodal models
```

> To enable seamless support for vision-language models in inference servers, your model needs to follow a few extra conventions on top of the general ones. These rules ensure that your model integrates properly with multimodal data.

Reviewer: ```suggestion
For multimodal models, you need to include a few more changes. The model must have a base `MyMultiModalModel` class, a nested config, and the `self.image_token` and `self.image_token_ids` attributes.
```

> 1. Your model must have a base `MyMultimodalModel` class that handles multimodal fusion without a language modeling head and a separate generative class that adds a head on top. The base model needs to implement a `get_image_features()` method that takes in image pixel values and returns encoded outputs. These will later be merged with language embeddings and thus should not require any postprocessing after. The shape of returned features has to match the number of input images. If the vision encoder returns variable-length outputs (e.g., patch-based), you can return a list of 2D tensors of size `(image_seq_len, image_dim)` - one per image. 

Reviewer: ```suggestion
1. A multimodal model requires a base `MyMultiModalModel` class to handle multimodal fusion without a language modeling head and a separate generative class that adds a head.

   The base model needs to implement the `get_image_features() method to accept image pixel values and return encoded outputs. These are later merged with the language embeddings and don't require any postprocessing. The shape of the returned features must match the number of input images. If a vision encoder returns variable-length outputs (patch-based), return a list of 2D tensors of size `(image_seq_len, image_dim)` for each image.

Expand the code below for an example.
```

<details>
<summary>modeling_my_multimodal_model.py</summary>

```python
from transformers.generation import GenerationMixin

class MyMultimodalModel(MyMultimodalPreTrainedModel):
    def __init__(self, config):
        super().__init__(config)
        self.language_model = AutoModel.from_config(config.text_config)
        self.vision_tower = AutoModel.from_config(config.vision_config)
        self.multimodal_projection = nn.Linear(vision_dim, text_dim)
    
    def get_image_features(self, pixel_values):
        return self.vision_tower(pixel_values).last_hidden_states
    

>     def forward(self, inpit_ids, pixel_values, **kwargs):

Reviewer: ```suggestion
    def forward(self, input_ids, pixel_values, **kwargs):
```

        # process your inputs
        return MyModelOutputWithPast(
            last_hidden_state=last_hidden_state,
            image_hidden_states=image_features,
            [...]
        )

class MyMultimodalModelForConditionalGeneration(MyMultimodalPreTrainedModel, GenerationMixin):
    def __init__(self, config):
        super().__init__(config)
        self.model = MyMultimodalModel(config)
        self.lm_head = nn.Linear(hidden_dim, vocab_size)
```
</details>

> 2. Your config must be nested, with a minimal set of the following fields:

Reviewer: ```suggestion
2. A multimodal model config must be nested with the following fields.
```

    * text_config: decoder language model config
    * vision_config: vision encoder config

>     * image_token_id: ID of the image placeholder token used in input to indicate image position

Reviewer: ```suggestion
    * image_token_id: ID of the image placeholder token used in the input to indicate image position
```

> 3. The model's processing class must have `self.image_token` and `self.image_token_ids` attributes. These are the placeholder tokens used to indicate image positions in the input. Note that it is the same token used by users when constructing a input prompt and the token that is used to masked scatter image features. Additionally, the class needs a `self._get_num_multimodal_tokens()` helper method that computes the number of placeholder tokens needed for multimodal inputs with given sizes and returns a `MultiModalData` object. Note that placeholder for row and column tokens are not counted as image placholders, only tokens that will actually be replaced by image features are computed.

Reviewer: ```suggestion
3. A multimodal model's processing class must have the `self.image_token` and `self.image_token_ids` attributes. These are placeholder tokens used to indicate image positions in the input. The placeholder token is the same token used in the input prompt and to mask scatter image features.

   The processing class also needs ` self._get_num_multimodal_tokens` method to compute the number of placeholder tokens needed for multimodal inputs with given sizes and to return a [`MultiModalData`] object. The placeholder for row and column tokens don't count as image placeholders. Only the tokens that are actually replaced by image features are computed.
```

> Finally, when `return_mm_token_type_ids=True`, the class has to return `mm_token_type_ids` indicating whether each position is a text token (`0`) or image placeholder token (`1`). Each image's token type IDs must be contiguous with no breaks between consecutive ones.

Reviewer: ```suggestion
Finally, when `return_mm_token_type_ids=True`, the class has to return `mm_token_type_ids` to indicate whether each position is a text token (`0`) or image placeholder token (`1`). Each image's token type IDs must be contiguous with no breaks between consecutive ones.

Expand the code below for an example.
```

            text_inputs["mm_token_type_ids"] = mm_token_type_ids.tolist()
        return BatchFeature(data={**text_inputs, **image_inputs}, tensor_type=return_tensors)

    def _get_num_multimodal_tokens(self, image_sizes=None, **kwargs):
        """
        Computes the number of placeholder tokens needed for multimodal inputs with the given sizes.
        Args:
            image_sizes (`list[list[int]]`, *optional*):
                The input sizes formatted as (height, width) per each image.
        Returns:
            `MultiModalData`: A `MultiModalData` object holding number of tokens per each of the provided
            input modalities, along with other useful data.
        """
        vision_data = {}
        if image_sizes is not None:
            num_image_tokens = [256] * len(image_sizes) # 256 placeholder tokens for each image always
            num_image_patches = [1] * len(image_sizes) # no patching, thus each image is processed as a single base image
            vision_data.update({"num_image_tokens": num_image_tokens, "num_image_patches": num_image_patches})
        return MultiModalData(**vision_data)
```
</details>

> ## Additional Resources

Reviewer: ```suggestion
## Resources
```

> Refer to the blog posts below to know more about how each inference server was integrated.
> 
> * [Blog Post in vLLM](https://blog.vllm.ai/2025/04/11/transformers-backend.html)
> * [Blog Post in SGLang](https://huggingface.co/blog/transformers-backend-sglang)

Reviewer: ```suggestion
* Read the [Transformers backend integration in vLLM](https://blog.vllm.ai/2025/04/11/transformers-backend.html) blog post for more details about the Transformers backend in vLLM.
* Read the [Transformers backend integration in SGLang](https://huggingface.co/blog/transformers-backend-sglang) blog post for more details about the Transformers backend in SGLang.
```
